csrc/fusion_segmenter.cpp

```diff
@@ -2643,6 +2595,70 @@ void deDuplicateScalarExprs(std::vector<Expr*>& exprs) {
 
 } // namespace
 
+std::vector<Expr*> SegmentedGroup::orderedExprs() const {
+  // The time complexity is O((V+E)LogV) where V is the number of nodes and E
+  // is the number of edges. LogV is due to the use of std::priority_queue to
+  // break ties by the original order.
+  std::unordered_map<Expr*, int64_t> original_order;
+  for (auto&& [i, e] : enumerate(exprs())) {
+    original_order[e] = i;
+  }
+
+  const std::unordered_set<Expr*> exprs_to_sort(exprs().begin(), exprs().end());
+
+  std::vector<Expr*> ordered_exprs;
+  ordered_exprs.reserve(exprs().size());
+
+  auto compare_by_original_order = [&original_order](Expr* a, Expr* b) {
+    // std::priority_queue implements a max heap, so this comparor returns true
+    // when RHS is originally ordered before LHS.
+    return original_order.at(a) > original_order.at(b);
+  };
+  std::priority_queue<
+      Expr*,
+      std::vector<Expr*>,
+      decltype(compare_by_original_order)>
+      to_visit(compare_by_original_order);
+
+  std::unordered_map<Expr*, int64_t> num_producers;
+  for (Expr* e : exprs()) {
+    int64_t& n = num_producers[e];
+    for (Val* in : e->inputs()) {
+      Expr* def = in->definition();
+      // Exprs in a SegmentedGroup come from the complete fusion, so the
+      // producer/consumer of an Expr may be outside the group. Therefore, we
+      // check exprs_to_sort.count.
+      if (exprs_to_sort.count(def) > 0) {
+        n++;
+      }
+    }
+
+    if (n == 0) {
+      to_visit.push(e);
+    }
+  }
+
+  while (!to_visit.empty()) {
+    Expr* e = to_visit.top();
+    to_visit.pop();
+
+    ordered_exprs.push_back(e);
```

Comment:
Wont this push expressions not in the group? to_visit from line 2637 will have expressions from the complete fusion whose producers are not in the group--meaning it can be from another group entirely. Am i misunderstanding the flow?

Reply:
There are only two `to_visit.push`es, line 2637 and line 2650. The `e` in line 2637 is in the group because `exprs()` returns only the group's Exprs. The `user` in line 2650 is also in the group because of the `exprs_to_sort.count(user)` check.

Reply:
Oh i see--I misread your comment as "`exprs()` in a SegmentedGroup come from the complete fusion...", as in the func returns expressions from the fusion. But`SegmentedGroup::exprs()` returning exprs in the group makes a lot more sense.